add RVV optimization for ZSTD_row_getMatchMask

w1m024 opened the pull request:
This pull request introduces a RISC-V Vector (RVV) specific optimization for the `ZSTD_row_getMatchMask` function, replacing the generic SWAR implementation on RV64 platforms with V-extension support. The goal is to leverage RVV's parallel computation capabilities to improve performance on the RISC-V architecture.

## Performance

### Microbenchmark Results

A microbenchmark isolating the `ZSTD_row_getMatchMask` function shows a significant speedup compared to the SWAR fallback.

| `rowEntries` | Speedup |
| :--- | :--- |
| 16 bytes | **5.87x** |
| 32 bytes | **9.63x** |
| 64 bytes | **17.98x** |

---

### Fullbench

The overall impact on the `fullbench` is modest. However, the new implementation shows a consistent small improvement and, most importantly, **no performance regression**.

## Validation

* All quick checks passed (`make check`).
* All long-running tests passed (`make test`).
* Static analysis reports no new issues (`make staticAnalyze`).

w1m024 committed:
add RVV optimization for ZSTD_row_getMatchMask

Co-authored-by: gong-flying <gongxiaofei24@iscas.ac.cn>

w1m024 commented:
Hi @camel-cdr @Cyan4973 ,
Following up on our previous discussion in PR #4478 , I have created this new pull request to address the feedback.
The new implementation resolves the portability issues by refactoring the code, as you suggested.
I would be grateful if you could take a look when you have a moment. Thanks again for your invaluable guidance!

camel-cdr commented:
Look, I'm not going to help you finetune your prompts until the llm generates a good implementation.

I've already toled you how to implement it earlier and probably wasted more time on this now then it would've taken me to write the implementation my self.

Cyan4973 commented on lib/compress/zstd_lazy.c:
```diff
@@ -1050,6 +1050,36 @@ ZSTD_row_getNEONMask(const U32 rowEntries, const BYTE* const src, const BYTE tag
+    int i;
+    assert(nbChunks == 1 || nbChunks == 2 || nbChunks == 4);
+
+    size_t vl = __riscv_vsetvl_e8m1(16);
```
minor: 
this is not `C90` compliant.

The `size_t vl` declaration should come _before_ the `assert()` expression.

Cyan4973 commented:
Is this new code path tested in CI ?
I know we have a `RISC-V` compilation test in CI,
but I am not sure about RVV.

w1m024 committed:
Refactor ZSTD_row_getMatchMask for RVV optimization

Performance (vs. SWAR)
- 16-byte data: 5.87x speedup
- 32-byte data: 9.63x speedup
- 64-byte data: 17.98x speedup

Co-authored-by: gong-flying <gongxiaofei24@iscas.ac.cn>

camel-cdr commented:
yes, this looks proper now 👍

w1m024 commented:
> Is this new code path tested in CI ? I know we have a `RISC-V` compilation test in CI, but I am not sure about RVV.

I've confirmed that the new RVV code path is covered 
The `dev-short-test.yml `file uses QEMU to run make check. It sets `CFLAGS="-march=rv64gcv"` and configures the virtual CPU with `-cpu rv64,v=true` and `vlen` values of 128, 256, and 512.
